controller/array_action/array_mediator_ds8k.py

```diff
@@ -370,17 +391,131 @@ def unmap_volume(self, volume_id, host_name):
         except exceptions.ClientException as ex:
             raise array_errors.UnMappingError(volume_id, host_name, ex.details)
 
+    def _get_pools(self):
+        logger.info("Getting pools")
+        try:
+            pools = self.client.get_pools()
+        except exceptions.ClientException as ex:
+            logger.error(
+                "Failed to get pools, reason is: {}".format(
+                    ex.details
+                )
+            )
+            raise ex
+        logger.info("Pools found: {}".format(pools))
+        return pools
+
+    def _get_api_volume_by_name(self, volume_name):
+        logger.info("Getting volume {}".format(volume_name))
+        pools = self._get_pools()
+        volume_candidates = []
+
+        for pool in pools:
+            volume_candidates.extend(self.client.get_volumes_by_pool(pool.id))
+
+        for volume in volume_candidates:
+            logger.info("Checking volume: {}".format(volume.name))
+            if volume.name == shorten_volume_name(volume_name, prefix=''):
+                logger.debug("Found volume: {}".format(volume))
+                return volume
+        return None
+
+    def _get_volume_by_name(self, volume_name):
+        api_volume = self._get_api_volume_by_name(volume_name)
+        return self._generate_volume_response(api_volume)
+
+    def _get_api_volume_by_id(self, volume_id):
+        try:
+            return self.client.get_volume(volume_id)
+        except exceptions.NotFound:
+            raise array_errors.VolumeNotFoundError(volume_id)
+
+    def _get_flashcopy(self, fc_id, not_exist_err=True):
+        logger.info("Getting flashcopy {}".format(fc_id))
+        try:
+            return self.client.get_flashcopies(fc_id)
+        except exceptions.NotFound as ex:
+            if ERROR_CODE_RESOURCE_NOT_EXISTS in str(ex.message).upper():
+                logger.info("{} not found".format(fc_id))
+                if not_exist_err:
+                    raise ex
+        except Exception as ex:
+            logger.exception(ex)
+            raise ex
+
     def get_snapshot(self, snapshot_name):
-        # TODO: CSI-1339
-        pass
+        logger.debug("Get snapshot : {}".format(snapshot_name))
+        candidate_api_volume = self._get_api_volume_by_name(snapshot_name)
+        if not candidate_api_volume:
+            return None
+        if not candidate_api_volume.flashcopy:
+            logger.error("FlashCopy Mapping not found for target volume: {}".format(snapshot_name))
+            raise array_errors.SnapshotNameBelongsToVolumeError(candidate_api_volume.name,
+                                                                self.service_address)
+        fcmap = self._get_flashcopy(candidate_api_volume.flashcopy[0].id)
+        source_volume_name = self.get_volume_name(fcmap.source_volume['id'])
```

Comment:
would `.source_volume.id` work?
if not, consider wrapping `get_cs_flashcopies` with a `Munch` in `RESTClient.get_flashcopies`

Reply:
it wont work and `Munch` is not "Munching" below first level, so it wont get to `source_volume` anyway

Reply:
I meant wrapping with `Munch.fromDict`

Reply:
done